commit d2ed04bdda089220e9e0d41582348cc4c71c7937
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-05-28T11:15:19+01:00 (4 months ago)
    
    TOOLS-2178 eng check targets should reference resources in deps/eng/tools
    Reviewed by: Trent Mick <trentm@gmail.com>
    Approved by: Trent Mick <trentm@gmail.com>

diff --git a/deps/eng b/deps/eng
index 126c0f0..d25b8fc 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit 126c0f032b8bfddd45b14e8ee14e73e9798a013f
+Subproject commit d25b8fc60fb8c649559924870fe3aaf75e7421d5
diff --git a/lib/cli.js b/lib/cli.js
index 023303b..ade7884 100644
--- a/lib/cli.js
+++ b/lib/cli.js
@@ -20,7 +20,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright (c) 2017, Joyent, Inc. All rights reserved.
+ * Copyright 2019 Joyent, Inc. All rights reserved.
  *
  * fwadm: CLI shared logic
  */
@@ -61,7 +61,7 @@ function displayRules(err, res, opts) {
         return;
     }
 
-    console.log('UUID                                 ENABLED RULE');
+    console.log('UUID                                 ENABLED LOG     RULE');
     res.forEach(function (r) {
         console.log(ruleLine(r));
     });
@@ -100,6 +100,11 @@ function getPayload(opts, args, callback) {
             payload.enabled = opts.enable;
         }
 
+
+        if (opts.log) {
+            payload.log = opts.log;
+        }
+
         ['description', 'global', 'owner_uuid'].forEach(function (p) {
             if (opts.hasOwnProperty(p)) {
                 payload[p] = opts[p];
@@ -198,8 +203,10 @@ function outputError(err, opts) {
  * Outputs one formatted rule line
  */
 function ruleLine(r) {
-    return util.format('%s %s %s', r.uuid,
-        r.enabled ? 'true   ' : 'false  ', r.rule);
+    return util.format('%s %s %s %s', r.uuid,
+        r.enabled ? 'true   ' : 'false  ',
+        r.log ? 'true   ' : 'false  ',
+        r.rule);
 }
 
 
diff --git a/lib/endpoints/rules/index.js b/lib/endpoints/rules/index.js
index ec25b83..8cbba2c 100644
--- a/lib/endpoints/rules/index.js
+++ b/lib/endpoints/rules/index.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -42,7 +42,8 @@ var LIST_SCHEMA = {
         action: validateAction,
         global: validate.boolean,
         tag: validateTag,
-        wildcard: validateArrayOfStrings
+        wildcard: validateArrayOfStrings,
+        log: validate.boolean
     }
 };
 
diff --git a/lib/fwapi-cli.js b/lib/fwapi-cli.js
index 91525a7..2528776 100644
--- a/lib/fwapi-cli.js
+++ b/lib/fwapi-cli.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
diff --git a/lib/rule.js b/lib/rule.js
index bae0bc8..26ccc6a 100644
--- a/lib/rule.js
+++ b/lib/rule.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2019, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -35,8 +35,9 @@ var tagUnescape = require('fwrule/lib/rule').tagUnescape;
  *
  * 1 - Initial version.
  * 2 - Save tag names/values unescaped, and index on "description".
+ * 3 - Add "log" field.
  */
-var MORAY_RAW_VERSION = 2;
+var MORAY_RAW_VERSION = 3;
 
 var BUCKET = {
     desc: 'fwrules',
@@ -56,6 +57,7 @@ var BUCKET = {
             'action': { 'type': 'string' },
             'protocol': { 'type': 'string' },
             'enabled': { 'type': 'boolean' },
+            'log': { 'type': 'boolean' },
             'fromwildcards': {
                 'type': '[string]'
             },
@@ -111,6 +113,8 @@ function ruleFromMoray(raw) {
     var data = {
         enabled: typeof (raw.enabled) === 'boolean' ? raw.enabled :
             (raw.enabled === 'true'),
+        log: typeof (raw.log) === 'boolean' ? raw.log :
+            (raw.log === 'true'),
         objectclass: raw.objectclass,
         parsed: {
             priority: raw.priority,
@@ -167,6 +171,11 @@ function ruleFromMoray(raw) {
         data.enabled = raw.enabled;
     }
 
+
+    if (typeof (raw.log) === 'boolean') {
+        data.log = raw.log;
+    }
+
     if (hasKey(raw, 'owner')) {
         data.owner_uuid = raw.owner;
     }
@@ -217,6 +226,8 @@ function ruleFromUFDS(raw) {
     var data = {
         enabled: typeof (raw.enabled) === 'boolean' ? raw.enabled :
             (raw.enabled === 'true'),
+        log: typeof (raw.log) === 'boolean' ? raw.log :
+            (raw.log === 'true'),
         objectclass: raw.objectclass,
         parsed: {
             action: raw.action,
@@ -259,6 +270,9 @@ function ruleFromUFDS(raw) {
         data.enabled = raw.enabled;
     }
 
+    if (typeof (raw.log) === 'boolean') {
+        data.log = raw.log;
+    }
     if (hasKey(raw, 'owner')) {
         data.owner_uuid = raw.owner;
     }
@@ -391,6 +405,7 @@ Rule.prototype.rawMoray = function _ruleRawMoray() {
     var raw = {
         action: this.action,
         enabled: this.enabled,
+        log: this.log,
         objectclass: Rule.objectclass,
         priority: this.priority,
         protocol: this.protocol,
@@ -494,6 +509,7 @@ Rule.prototype.rawUFDS = function _ruleRawUFDS() {
     var raw = {
         action: this.action,
         enabled: this.enabled,
+        log: this.log,
         objectclass: Rule.objectclass,
         protocol: this.protocol,
         uuid: this.uuid,
diff --git a/lib/ufds/filter.js b/lib/ufds/filter.js
index b9ce76c..0634ec5 100644
--- a/lib/ufds/filter.js
+++ b/lib/ufds/filter.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -95,6 +95,15 @@ function ruleCommonFilter(opts) {
         filter.push(eq('enabled', params.enabled));
     }
 
+
+    if (hasKey(params, 'log')) {
+        if ((params.log !== 'true') && (params.log !== 'false')) {
+            throw new restify.InvalidArgumentError(
+                'Invalid value for log: must be true or false');
+        }
+        filter.push(eq('log', params.log));
+    }
+
     if (hasKey(params, 'port')) {
         if (!validators.validatePortOrAll(params.port)) {
             throw new restify.InvalidArgumentError(
@@ -151,6 +160,7 @@ function ruleCommonFilter(opts) {
  *     Optional parameters:
  *     - action {String}
  *     - enabled {Boolean}
+ *     - log {Boolean}
  *     - port {Integer}
  *     - protocol {String}
  *     - ip {String or Array}
@@ -294,6 +304,7 @@ function ruleMorayFilter(opts) {
  *     Optional parameters:
  *     - action {String}
  *     - enabled {Boolean}
+ *     - log {Boolean}
  *     - port {Integer}
  *     - protocol {String}
  *     - ip {String or Array}
diff --git a/package.json b/package.json
index 0c47f80..6555450 100644
--- a/package.json
+++ b/package.json
@@ -20,7 +20,7 @@
     "effluent-logger": "git+https://github.com/joshwilsdon/effluent-logger.git#d662f161a07f94045ad2afb45442931511c40e51",
     "extsprintf": "1.0.2",
     "fast-messages": "1.0.1",
-    "fwrule": "2.0.0",
+    "fwrule": "2.1.0",
     "jsprim": "^2.0.0",
     "moray": "^3.5.0",
     "moray-filter": "1.0.0",
diff --git a/test/integration/get.test.js b/test/integration/get.test.js
index e782686..a1348db 100644
--- a/test/integration/get.test.js
+++ b/test/integration/get.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -37,17 +37,20 @@ var VM_UUIDS = [
 var RULES = [
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM any TO all vms ALLOW tcp PORT 5000'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[1],
         rule: 'FROM (tag "foo" = "bar" OR tag "foo" = "baz") '
             + 'TO tag "side" = "two" ALLOW tcp (PORT 5003 AND PORT 5004)'
     },
     {
         enabled: true,
+        log: false,
         global: true,
         rule: 'FROM any TO tag "foo" = "baz" ALLOW tcp PORT 5010'
     },
@@ -55,32 +58,38 @@ var RULES = [
     /* IP rule, VM rule, subnet rule */
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM ip 8.8.8.8 TO tag "foo" = "baz" ALLOW tcp PORT 5010'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM ip 4.4.4.4 TO tag "foo" = "baz" ALLOW tcp PORT 5010'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM subnet 10.8.0.0/16 TO tag "foo" = "baz" ALLOW tcp PORT 5010'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM subnet 10.7.0.0/16 TO tag "foo" = "baz" ALLOW tcp PORT 5010'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM vm ' + VM_UUIDS[0] +
             ' TO tag "foo" = "baz" ALLOW tcp PORT 5010'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM vm ' + VM_UUIDS[1] +
             ' TO tag "foo" = "baz" ALLOW tcp PORT 5010'
@@ -89,16 +98,19 @@ var RULES = [
     /* FWRULE_VERSION 4 features */
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[2],
         rule: 'FROM tag "a" TO tag "b" ALLOW tcp PORT 80 PRIORITY 50'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[2],
         rule: 'FROM tag "a" TO tag "b" ALLOW ah'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[2],
         rule: 'FROM tag "a" TO tag "b" ALLOW esp'
     }
diff --git a/test/integration/global.test.js b/test/integration/global.test.js
index 2843d1e..0baccce 100644
--- a/test/integration/global.test.js
+++ b/test/integration/global.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright 2019, Joyent, Inc.
  */
 
 /*
@@ -46,6 +46,7 @@ test('Add rule', function (t) {
     RULES.push({
         enabled: true,
         global: true,
+        log: false,
         rule: 'FROM any TO all vms ALLOW udp PORT 5000'
     });
 
diff --git a/test/integration/invalid.test.js b/test/integration/invalid.test.js
index 932669e..216c43e 100644
--- a/test/integration/invalid.test.js
+++ b/test/integration/invalid.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -60,6 +60,15 @@ function createPayload(ruleTxt) {
       },
       [ mod_err.invalidParam('global', 'global must be true or false') ] ],
 
+    [ 'bad log value',
+      {
+          rule: 'FROM tag a TO tag b ALLOW udp port 53',
+          enabled: true,
+          global: true,
+          log: 'foobar'
+      },
+      [ mod_err.invalidParam('log', 'log must be true or false') ] ],
+
     [ 'bad IPv4 subnet: bits to right of mask',
       createPayload('FROM tag foo TO subnet 10.8.0.0/5 ALLOW udp port 53'),
       [ mod_err.invalidParam('rule',
diff --git a/test/integration/list.test.js b/test/integration/list.test.js
index a40a441..0c2d79f 100644
--- a/test/integration/list.test.js
+++ b/test/integration/list.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2019, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -36,21 +36,25 @@ var RULES = [
     // Add another rule for this user to make sure we're not return all
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM any TO all vms ALLOW tcp PORT 5000'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM all vms TO subnet 10.2.1.0/24 ALLOW tcp PORT 5001'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM subnet 10.2.1.0/24 TO all vms ALLOW tcp PORT 5002'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM subnet 10.3.1.0/24 TO all vms ALLOW tcp PORT 5002'
     },
@@ -59,16 +63,19 @@ var RULES = [
 
     {
         enabled: true,
+        log: true,
         owner_uuid: OWNERS[1],
         rule: 'FROM subnet 10.2.1.0/24 TO all vms ALLOW tcp PORT 5001'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[1],
         rule: 'FROM tag "foo" TO all vms BLOCK udp PORT all'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[1],
         rule: 'FROM (tag "foo" = "bar" OR tag "foo" = "baz") '
             + 'TO tag "side" = "two" ALLOW tcp (PORT 5003 AND PORT 5004)'
@@ -78,26 +85,31 @@ var RULES = [
 
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[2],
         rule: 'FROM any TO all vms ALLOW ah'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[2],
         rule: 'FROM any TO all vms ALLOW esp'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[2],
         rule: 'FROM any TO tag "a" = "q" ALLOW icmp TYPE all'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[2],
         rule: 'FROM any TO tag "b" ALLOW icmp6 TYPE all'
     },
     {
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[2],
         rule: 'FROM tag "c" TO any BLOCK udp PORT 53'
     }
@@ -143,6 +155,31 @@ test('list: subnet 10.3.1.0/24', function (t) {
     });
 });
 
+test('list: log true', function (t) {
+    mod_rule.list(t, {
+        params: {
+            owner_uuid: OWNERS[1],
+            log: true
+        },
+        exp: [
+            RULES[4]
+        ]
+    });
+});
+
+
+test('list: log false', function (t) {
+    mod_rule.list(t, {
+        params: {
+            owner_uuid: OWNERS[1],
+            log: false
+        },
+        exp: [
+            RULES[5], RULES[6]
+        ]
+    });
+});
+
 test('list: all ports', function (t) {
     mod_rule.createAndGet(t, {
         rule: {
@@ -152,6 +189,7 @@ test('list: all ports', function (t) {
         },
         exp: {
             enabled: true,
+            log: false,
             owner_uuid: OWNERS[3],
             rule: 'FROM tag "foo" TO all vms BLOCK udp PORT all'
         }
diff --git a/test/integration/provision.multi-test.js b/test/integration/provision.multi-test.js
index 441b97b..9896a9a 100644
--- a/test/integration/provision.multi-test.js
+++ b/test/integration/provision.multi-test.js
@@ -83,6 +83,7 @@ test('Add rules', OPTS, function (t) {
         RULES.ssh = {
             description: 'allow SSH',
             enabled: true,
+            log: false,
             owner_uuid: OWNERS[0],
             rule: util.format('FROM any TO tag "%s" ALLOW tcp PORT 22',
                 TAGS.ssh)
@@ -98,6 +99,7 @@ test('Add rules', OPTS, function (t) {
         RULES.dns = {
             description: 'allow DNS',
             enabled: true,
+            log: false,
             owner_uuid: OWNERS[0],
             rule: util.format('FROM any TO tag "%s" ALLOW udp PORT 53',
                 TAGS.dns)
@@ -227,6 +229,7 @@ test('Add disabled rule', OPTS, function (t) {
         RULES.db = {
             description: 'allow DB',
             enabled: false,
+            log: false,
             owner_uuid: OWNERS[0],
             rule: util.format('FROM (tag "%s" = "1" OR tag "%s" = "2") TO ' +
                 '(tag "%s" = "1" OR tag "%s" = "2") ALLOW tcp PORT 5432',
@@ -387,6 +390,7 @@ test('Add VMs rule', OPTS, function (t) {
     t.test('add', function (t2) {
         RULES.https = {
             enabled: true,
+            log: false,
             owner_uuid: OWNERS[0],
             rule: util.format('FROM vm %s TO vm %s ALLOW tcp PORT 443',
                 VMS[2].uuid, VMS[1].uuid)
diff --git a/test/integration/resolve.test.js b/test/integration/resolve.test.js
index c73a4b9..1418943 100644
--- a/test/integration/resolve.test.js
+++ b/test/integration/resolve.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2019, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -182,7 +182,8 @@ test('setup', function (t) {
                 owner_uuid: OWNERS[1],
                 rule: fmt('FROM vm %s TO all vms BLOCK udp PORT 54',
                     VMS[2]),
-                enabled: true
+                enabled: true,
+                log: false
             }
         },
 
@@ -190,7 +191,8 @@ test('setup', function (t) {
             oneToAll: {
                 owner_uuid: OWNERS[2],
                 rule: 'FROM tag "one" TO all vms BLOCK udp PORT 55',
-                enabled: true
+                enabled: true,
+                log: false
             }
         },
 
@@ -198,7 +200,8 @@ test('setup', function (t) {
             oneToAll: {
                 owner_uuid: OWNERS[3],
                 rule: 'FROM tag "one" TO all vms ALLOW udp PORT 56',
-                enabled: true
+                enabled: true,
+                log: false
             }
         },
 
@@ -206,7 +209,8 @@ test('setup', function (t) {
             allToOne: {
                 owner_uuid: OWNERS[4],
                 rule: 'FROM all vms TO tag "one" BLOCK udp PORT 57',
-                enabled: true
+                enabled: true,
+                log: false
             }
         },
 
@@ -214,7 +218,8 @@ test('setup', function (t) {
             allToOne: {
                 owner_uuid: OWNERS[5],
                 rule: 'FROM all vms TO tag "one" ALLOW udp PORT 58',
-                enabled: true
+                enabled: true,
+                log: false
             }
         },
 
@@ -222,20 +227,23 @@ test('setup', function (t) {
             vmToOne: {
                 owner_uuid: OWNERS[6],
                 rule: fmt('FROM vm %s TO tag "one" ALLOW udp PORT 59', VMS[4]),
-                enabled: true
+                enabled: true,
+                log: false
             },
 
             vmToOneTwo: {
                 owner_uuid: OWNERS[6],
                 rule: fmt('FROM vm %s TO tag "one" = "two" ALLOW udp PORT 59',
                     VMS[4]),
-                enabled: true
+                enabled: true,
+                log: false
             },
 
             vmToOneThree: {
                 owner_uuid: OWNERS[6],
                 rule: 'FROM any TO tag "one" = "three" ALLOW udp PORT 59',
-                enabled: true
+                enabled: true,
+                log: false
             },
 
             vmToMultiTags: {
@@ -243,21 +251,24 @@ test('setup', function (t) {
                 rule: fmt('FROM vm %s TO '
                     + '(tag "five" = "six" OR tag "three" = "four") '
                     + 'ALLOW udp PORT 58', VMS[3]),
-                enabled: true
+                enabled: true,
+                log: false
             },
 
             ipToVm5: {
                 owner_uuid: OWNERS[6],
                 rule: fmt('FROM ip 10.1.2.5 TO vm %s ALLOW tcp PORT 80',
                     VMS[5]),
-                enabled: true
+                enabled: true,
+                log: false
             },
 
             vm6ToIp: {
                 owner_uuid: OWNERS[6],
                 rule: fmt('FROM vm %s TO ip 10.1.2.5 BLOCK tcp PORT 80',
                     VMS[6]),
-                enabled: true
+                enabled: true,
+                log: false
             },
 
             vmsOnBothSides: {
@@ -265,7 +276,8 @@ test('setup', function (t) {
                 rule: fmt('FROM (ip 10.1.2.5 OR vm %s OR vm %s) TO '
                     + '(ip 10.1.2.5 OR vm %s OR vm %s) ALLOW udp PORT 5432',
                     VMS[5], VMS[6], VMS[5], VMS[6]),
-                enabled: true
+                enabled: true,
+                log: false
             }
         }
     };
@@ -273,6 +285,7 @@ test('setup', function (t) {
     for (r in RULES.o0) {
         RULES.o0[r] = {
             enabled: true,
+            log: false,
             owner_uuid: OWNERS[0],
             rule: RULES.o0[r].rule
         };
diff --git a/test/integration/update.test.js b/test/integration/update.test.js
index cb226c4..c79bde0 100644
--- a/test/integration/update.test.js
+++ b/test/integration/update.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -39,6 +39,7 @@ var RULES = [];
 test('Add rule', function (t) {
     RULES.push({
         enabled: true,
+        log: false,
         owner_uuid: OWNERS[0],
         rule: 'FROM all vms TO tag "test" BLOCK tcp PORT 8000'
     });
diff --git a/test/integration/vm-update.multi-test.js b/test/integration/vm-update.multi-test.js
index 414ecfa..e83c177 100644
--- a/test/integration/vm-update.multi-test.js
+++ b/test/integration/vm-update.multi-test.js
@@ -80,6 +80,7 @@ test('Add rules', OPTS, function (t) {
         RULES.ssh1 = {
             description: 'allow SSH',
             enabled: true,
+            log: false,
             owner_uuid: OWNERS[0],
             rule: util.format(
                 'FROM tag "%s" = "one" TO tag "%s" = "two" ALLOW tcp PORT 22',
@@ -96,6 +97,7 @@ test('Add rules', OPTS, function (t) {
         RULES.web1 = {
             description: 'allow DB access',
             enabled: true,
+            log: false,
             owner_uuid: OWNERS[0],
             rule: util.format(
                 'FROM tag "%s" TO tag "%s" ALLOW tcp PORT 5432',
diff --git a/test/runtest b/test/runtest
index f386a9f..99d18d7 100755
--- a/test/runtest
+++ b/test/runtest
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 #
 
 #
@@ -14,7 +14,6 @@
 
 TOP=$(cd $(dirname $0)/../; pwd)
 NODE_INSTALL=$TOP/node
-NODEUNIT=$TOP/node_modules/.bin/nodeunit
 UNAME=$(uname -s)
 
 guard_file=/lib/sdc/.sdc-test-no-production-data
diff --git a/test/unit/data/migration.js b/test/unit/data/migration.js
index 8011149..0ba7986 100644
--- a/test/unit/data/migration.js
+++ b/test/unit/data/migration.js
@@ -166,6 +166,7 @@ rules[RULE_1_UUID] = {
         'uuid': RULE_1_UUID,
         'version': RULE_1_VERSION,
         'owner_uuid': OWNER_1,
+        'log': false,
         'enabled': true
     }
 };
@@ -208,6 +209,7 @@ rules[RULE_2_UUID] = {
         'uuid': RULE_2_UUID,
         'version': RULE_2_VERSION,
         'owner_uuid': OWNER_1,
+        'log': false,
         'enabled': true
     }
 };
@@ -253,6 +255,7 @@ rules[RULE_3_UUID] = {
         'uuid': RULE_3_UUID,
         'version': RULE_3_VERSION,
         'owner_uuid': OWNER_2,
+        'log': false,
         'enabled': false
     }
 };
@@ -293,6 +296,7 @@ rules[RULE_4_UUID] = {
         'uuid': RULE_4_UUID,
         'version': RULE_4_VERSION,
         'owner_uuid': OWNER_2,
+        'log': false,
         'enabled': false
     }
 };
@@ -331,6 +335,7 @@ rules[RULE_5_UUID] = {
         'uuid': RULE_5_UUID,
         'version': RULE_5_VERSION,
         'enabled': true,
+        'log': false,
         'global': true
     }
 };
@@ -375,6 +380,7 @@ rules[RULE_6_UUID] = {
         'uuid': RULE_6_UUID,
         'version': RULE_6_VERSION,
         'owner_uuid': OWNER_3,
+        'log': false,
         'enabled': true
     }
 };
@@ -414,6 +420,7 @@ rules[RULE_7_UUID] = {
         'uuid': RULE_7_UUID,
         'version': RULE_7_VERSION,
         'owner_uuid': OWNER_4,
+        'log': false,
         'enabled': true
     }
 };
@@ -453,6 +460,7 @@ rules[RULE_8_UUID] = {
         'uuid': RULE_8_UUID,
         'version': RULE_8_VERSION,
         'owner_uuid': OWNER_4,
+        'log': false,
         'enabled': true
     }
 };
@@ -491,6 +499,7 @@ rules[RULE_9_UUID] = {
         'uuid': RULE_9_UUID,
         'version': RULE_9_VERSION,
         'owner_uuid': OWNER_4,
+        'log': false,
         'enabled': true
     }
 };
@@ -516,6 +525,7 @@ rules[RULE_10_UUID] = {
         'uuid': RULE_10_UUID,
         'version': RULE_10_VERSION,
         'enabled': true,
+        'log': false,
         'global': true
     }
 };
@@ -540,6 +550,7 @@ rules[RULE_11_UUID] = {
         'uuid': RULE_11_UUID,
         'owner_uuid': OWNER_4,
         'version': RULE_11_VERSION,
+        'log': false,
         'enabled': true
     }
 };
@@ -565,6 +576,7 @@ rules[RULE_12_UUID] = {
         'uuid': RULE_12_UUID,
         'owner_uuid': OWNER_4,
         'version': RULE_12_VERSION,
+        'log': false,
         'enabled': true
     }
 };
@@ -618,6 +630,7 @@ function generateRule() {
             'uuid': rule_uuid,
             'version': version,
             'owner_uuid': owner_uuid,
+            'log': false,
             'enabled': true
         }
     };
diff --git a/test/unit/rule.test.js b/test/unit/rule.test.js
index f4dc25a..3c3ec85 100644
--- a/test/unit/rule.test.js
+++ b/test/unit/rule.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -83,6 +83,7 @@ test('all target types', function (t) {
         tovm: [vms[1]],
 
         enabled: true,
+        log: false,
         objectclass: mod_rule.objectclass,
         ports: [ 80, 81 ],
         action: 'allow',
@@ -95,6 +96,8 @@ test('all target types', function (t) {
     inRule.version = rule.version;
 
     t.deepEqual(rule.rawUFDS(), raw, 'rule.rawUFDS()');
+
+    inRule.log = rule.log;
     t.deepEqual(rule.serialize(), inRule, 'rule.serialize()');
     t.equal(rule.dn, util.format('uuid=%s, ou=fwrules, o=smartdc', rule.uuid),
         'rule.dn');
@@ -125,6 +128,7 @@ test('owner_uuid', function (t) {
         fromip: [ util_ip.aton(ip) ],
         tovm: [ vm ],
         enabled: true,
+        log: false,
         objectclass: mod_rule.objectclass,
         owner: inRule.owner_uuid,
         ports: [ 25 ],
@@ -138,6 +142,8 @@ test('owner_uuid', function (t) {
     inRule.version = rule.version;
 
     t.deepEqual(rule.rawUFDS(), raw, 'rule.rawUFDS()');
+
+    inRule.log = rule.log;
     t.deepEqual(rule.serialize(), inRule, 'rule.serialize()');
     t.equal(rule.dn, util.format('uuid=%s, ou=fwrules, o=smartdc', rule.uuid),
         'rule.dn');
@@ -165,6 +171,7 @@ test('multiple tags with multiple quoted values', function (t) {
 
     var raw = {
         enabled: false,
+        log: false,
         objectclass: mod_rule.objectclass,
         ports: [ 80 ],
         action: 'allow',
@@ -187,7 +194,8 @@ test('multiple tags with multiple quoted values', function (t) {
             + '(tag "some tag" = "value" OR tag "some-tag" = "another value") '
             + 'ALLOW tcp PORT 80',
         uuid: rule.uuid,
-        version: rule.version
+        version: rule.version,
+        log: false
     };
     t.deepEqual(rule.serialize(), serialized, 'rule.serialize()');
 
@@ -204,7 +212,6 @@ test('multiple tags with multiple quoted values', function (t) {
     // Now check that we can reconstruct this data from UFDS
     rule = mod_rule.create(raw, app);
     t.deepEqual(rule.rawUFDS(), raw, 'rule.rawUFDS()');
-    t.deepEqual(rule.rawUFDS(), raw, 'rule.rawUFDS()');
     t.ok(!rule.allVMs, 'rule.allVMs');
     t.deepEqual(rule.tags, ruleTags, 'rule.tags');
 
diff --git a/test/unit/ufds-smoke-test.test.js b/test/unit/ufds-smoke-test.test.js
index 8cf314a..22c4d2b 100644
--- a/test/unit/ufds-smoke-test.test.js
+++ b/test/unit/ufds-smoke-test.test.js
@@ -131,6 +131,7 @@ test('Ping FWAPI', function (t) {
 test('Add rule 1', function (t) {
     RULES.push({
         enabled: true,
+        log: false,
         owner_uuid: VMS[0].owner_uuid,
         rule: util.format('FROM vm %s TO vm %s ALLOW tcp PORT 80',
             VMS[0].uuid, VMS[1].uuid)
@@ -176,6 +177,7 @@ test('Add rule 1', function (t) {
 test('Add rule 2', function (t) {
     RULES.push({
         enabled: true,
+        log: false,
         owner_uuid: VMS[0].owner_uuid,
         rule: 'FROM tag "foo" TO tag "bar" = "\\)" ALLOW tcp PORT 80'
     });
diff --git a/test/unit/ufds-to-moray.test.js b/test/unit/ufds-to-moray.test.js
index 609f27a..94cd146 100644
--- a/test/unit/ufds-to-moray.test.js
+++ b/test/unit/ufds-to-moray.test.js
@@ -233,6 +233,7 @@ test('New rule goes to Moray bucket', function (t) {
         rule: 'FROM tag "foo" = "hello-world" TO tag "bar" ALLOW tcp PORT 22',
         description: 'Rule added after migration',
         enabled: true,
+        log: false,
         owner_uuid: migr_data.OWNER_4
     };
 
diff --git a/test/unit/update.test.js b/test/unit/update.test.js
index 1232c5c..88a300f 100644
--- a/test/unit/update.test.js
+++ b/test/unit/update.test.js
@@ -55,6 +55,7 @@ test('setup', function (t) {
 test('Add rule', function (t) {
     RULES.push({
         enabled: true,
+        log: false,
         owner_uuid: VMS[0].owner_uuid,
         rule: util.format('FROM vm %s TO vm %s ALLOW tcp PORT 80',
             VMS[0].uuid, VMS[1].uuid)
diff --git a/test/unit/v1-to-latest.test.js b/test/unit/v1-to-latest.test.js
index 9eb0825..8275ede 100644
--- a/test/unit/v1-to-latest.test.js
+++ b/test/unit/v1-to-latest.test.js
@@ -238,6 +238,7 @@ test('No new v1 rules created', function (t) {
         rule: 'FROM tag "foo" = "hello-world" TO tag "bar" ALLOW tcp PORT 22',
         description: 'Rule added after migration',
         enabled: true,
+        log: false,
         owner_uuid: migr_data.OWNER_4
     };
 
diff --git a/tools/bashstyle b/tools/bashstyle
deleted file mode 100755
index 38a287c..0000000
--- a/tools/bashstyle
+++ /dev/null
@@ -1,157 +0,0 @@
-#!/usr/bin/env node
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
-/*
- * Copyright (c) 2014, Joyent, Inc.
- */
-
-/*
- * bashstyle: check bash scripts for adherence to style guidelines, including:
- *
- *    o no lines longer than 80 characters
- *    o file does not end with a blank line
- *    o Do not use 'local' and var initialization *using a subshell* in the
- *      same statement. See
- *      <http://www.tldp.org/LDP/abs/html/localvar.html#EXITVALANOMALY01>
- *      for why not. Arguably this belongs in a separate 'bashlint'.
- *
- * Future enhancements could include:
- *    o indents consistent with respect to tabs, spaces
- *    o indents consistently sized (all are some multiple of the smallest
- *      indent, which must be a tab or 4 or 8 spaces)
- */
-
-var VERSION = '2.0.0';
-
-var mod_assert = require('assert');
-var mod_fs = require('fs');
-
-var nerrors = 0;
-
-main();
-process.exit(0);
-
-function main()
-{
-	var files = process.argv.slice(2);
-
-	if (files.length === 0) {
-		console.error('usage: %s file1 [...]',
-		    process.argv.slice(0, 2).join(' '));
-		process.exit(2);
-	}
-
-	files.forEach(checkFile);
-
-	if (nerrors != 0)
-		process.exit(1);
-}
-
-function checkFile(filename)
-{
-	var text = mod_fs.readFileSync(filename, 'utf-8');
-	var lines = text.split('\n');
-	var i;
-	var styled = false;
-	var styleStart;
-
-	mod_assert.ok(lines.length > 0);
-
-	/*
-	 * Expand tabs in each line and check for long lines.
-	 */
-	for (i = 1; i <= lines.length; i++) {
-		var line = expandTabs(lines[i - 1]);
-
-		if (i > 1 && lines[i-2].match(/# BASHSTYLED/)) {
-			continue;
-		}
-
-		if (line.match(/# BEGIN BASHSTYLED/)) {
-			styleStart = i;
-			styled = true;
-		}
-
-		if (line.match(/# END BASHSTYLED/)) {
-			if (styled != true) {
-				nerrors++;
-				console.log('%s: %d: END BASHSTYLED w/o corresponding BEGIN',
-				            filename, i);
-			}
-			styled = false;
-		}
-
-		if (!styled && line.match(/^\s*local\s+(\w+)\s*=.*\$\(/)) {
-			nerrors++;
-			var m = line.match(/^\s*local\s+(\w+)\s*=/);
-			console.log('%s: %d: declaring and setting a "local" ' +
-				'var in the same statement ' +
-				'ignores a subshell return code ' +
-				'<http://www.tldp.org/LDP/abs/html/localvar.html#EXITVALANOMALY01>: ' +
-				'local %s=...',
-				filename, i, m[1]);
-		}
-
-		if (!styled && line.length > 80) {
-			nerrors++;
-			console.log('%s: %d: line exceeds 80 columns',
-			    filename, i);
-		}
-	}
-
-	if (styled) {
-		nerrors++;
-		console.log('%s: %d: BEGIN BASHSTYLED that does not END',
-		            filename, styleStart);
-	}
-
-
-	/*
-	 * No sane editor lets you save a file without a newline at the very end.
-	 */
-	if (lines[lines.length - 1].length !== 0) {
-		nerrors++;
-		console.log('%s: %d: file does not end with newline',
-			filename, lines.length);
-	}
-
-	/*
-	 * Since the file will always end with a newline, the last entry of
-	 * "lines" will actually be blank.
-	 */
-	if (lines.length > 1 && lines[lines.length - 2].length === 0) {
-		nerrors++;
-		console.log('%s: %d: file ends with a blank line',
-		    filename, lines.length - 1);
-	}
-}
-
-function expandTabs(text)
-{
-	var out = '';
-	var col = 0;
-	var j, k;
-
-	for (j = 0; j < text.length; j++) {
-		if (text[j] != '\t') {
-			out += text[j];
-			col++;
-			continue;
-		}
-
-		k = 8 - (col % 8);
-		col += k;
-
-		do {
-			out += ' ';
-		}  while (--k > 0);
-
-		col += k;
-	}
-
-	return (out);
-}
diff --git a/tools/service_bundle.dtd.1 b/tools/service_bundle.dtd.1
deleted file mode 100644
index e5c2380..0000000
--- a/tools/service_bundle.dtd.1
+++ /dev/null
@@ -1,1091 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<!--
- Copyright (c) 2004, 2010, Oracle and/or its affiliates. All rights reserved.
-
- CDDL HEADER START
-
- The contents of this file are subject to the terms of the
- Common Development and Distribution License (the "License").
- You may not use this file except in compliance with the License.
-
- You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
- or http://www.opensolaris.org/os/licensing.
- See the License for the specific language governing permissions
- and limitations under the License.
-
- When distributing Covered Code, include this CDDL HEADER in each
- file and include the License file at usr/src/OPENSOLARIS.LICENSE.
- If applicable, add the following below this CDDL HEADER, with the
- fields enclosed by brackets "[]" replaced with your own identifying
- information: Portions Copyright [yyyy] [name of copyright owner]
-
- CDDL HEADER END
--->
-
-<!--
-  Service description DTD
-
-    Most attributes are string values (or an individual string from a
-    restricted set), but attributes with a specific type requirement are
-    noted in the comment describing the element.
--->
-
-<!--
-  XInclude support
-
-    A series of service bundles may be composed via the xi:include tag.
-    smf(5) tools enforce that all bundles be of the same type.
--->
-
-<!--
-     These entities are used for the property, propval and property_group
-     elements, that require type attributes for manifest, while for profiles
-     the type attributes are only implied.
--->
-
-<!ENTITY % profile "IGNORE">
-<!ENTITY % manifest "INCLUDE">
-
-<!ELEMENT xi:include
-  (xi:fallback)
-  >
-<!ATTLIST xi:include
-  href CDATA #REQUIRED
-  parse (xml|text) "xml"
-  encoding CDATA #IMPLIED
-  xmlns:xi CDATA #FIXED "http://www.w3.org/2001/XInclude"
-  >
-
-<!ELEMENT xi:fallback
-  ANY
-  >
-<!ATTLIST xi:fallback
-  xmlns:xi CDATA #FIXED "http://www.w3.org/2001/XInclude"
-  >
-
-<!--
-  stability
-
-    This element associates an SMI stability level with the parent
-    element.  See attributes(5) for an explanation of interface
-    stability levels.
-
-    Its attribute is
-
-	value	The stability level of the parent element.
--->
-
-<!ELEMENT stability EMPTY>
-
-<!ATTLIST stability
-	value		( Standard | Stable | Evolving | Unstable |
-			External | Obsolete ) #REQUIRED >
-
-<!-- Property value lists -->
-
-<!--
-  value_node
-
-    This element represents a single value within any of the typed
-    property value lists.
-
-    Its attribute is
-
-	value	The value for this node in the list.
--->
-
-<!ELEMENT value_node EMPTY>
-
-<!ATTLIST value_node
-	value CDATA #REQUIRED>
-
-<!--
-  count_list
-  integer_list
-  opaque_list
-  host_list
-  hostname_list
-  net_address_list
-  net_address_v4_list
-  net_address_v6_list
-  time_list
-  astring_list
-  ustring_list
-  boolean_list
-  fmri_list
-  uri_list
-
-    These elements represent the typed lists of values for a property.
-    Each contains one or more value_node elements representing each
-    value on the list.
-
-    None of these elements has attributes.
--->
-
-<!ELEMENT count_list
-	( value_node+ )>
-
-<!ATTLIST count_list>
-
-<!ELEMENT integer_list
-	( value_node+ )>
-
-<!ATTLIST integer_list>
-
-<!ELEMENT opaque_list
-	( value_node+ )>
-
-<!ATTLIST opaque_list>
-
-<!ELEMENT host_list
-	( value_node+ )>
-
-<!ATTLIST host_list>
-
-<!ELEMENT hostname_list
-	( value_node+ )>
-
-<!ATTLIST hostname_list>
-
-<!ELEMENT net_address_list
-	( value_node+ )>
-
-<!ATTLIST net_address_list>
-
-<!ELEMENT net_address_v4_list
-	( value_node+ )>
-
-<!ATTLIST net_address_v4_list>
-
-<!ELEMENT net_address_v6_list
-	( value_node+ )>
-
-<!ATTLIST net_address_v6_list>
-
-<!ELEMENT time_list
-	( value_node+ )>
-
-<!ATTLIST time_list>
-
-<!ELEMENT astring_list
-	( value_node+ )>
-
-<!ATTLIST astring_list>
-
-<!ELEMENT ustring_list
-	( value_node+ )>
-
-<!ATTLIST ustring_list>
-
-<!ELEMENT boolean_list
-	( value_node+ )>
-
-<!ATTLIST boolean_list>
-
-<!ELEMENT fmri_list
-	( value_node+ )>
-
-<!ATTLIST fmri_list>
-
-<!ELEMENT uri_list
-	( value_node+ )>
-
-<!ATTLIST uri_list>
-
-<!-- Properties and property groups -->
-
-<!--
-   property
-
-     This element is for a singly or multiply valued property within a
-     property group.  It contains an appropriate value list element,
-     which is expected to be consistent with the type attribute.
-
-     Its attributes are
-
-	name	The name of this property.
-
-	type	The data type for this property.
-
-	override These values should replace values already in the
-		repository.
--->
-
-<![%profile;[
-<!ELEMENT property
-	( count_list | integer_list | opaque_list | host_list | hostname_list |
-	net_address_list | net_address_v4_list | net_address_v6_list |
-	time_list | astring_list | ustring_list | boolean_list | fmri_list |
-	uri_list )? >
-
-<!ATTLIST property
-	name		CDATA #REQUIRED
-	type		( count | integer | opaque | host | hostname |
-			net_address | net_address_v4 | net_address_v6 | time |
-			astring | ustring | boolean | fmri | uri ) #IMPLIED
-	override	( true | false ) "false" >
-]]>
-	
-<![%manifest;[
-<!ELEMENT property
-	( count_list | integer_list | opaque_list | host_list | hostname_list |
-	net_address_list | net_address_v4_list | net_address_v6_list |
-	time_list | astring_list | ustring_list | boolean_list | fmri_list |
-	uri_list )? >
-
-<!ATTLIST property
-	name		CDATA #REQUIRED
-	type		( count | integer | opaque | host | hostname |
-			net_address | net_address_v4 | net_address_v6 | time |
-			astring | ustring | boolean | fmri | uri ) #REQUIRED
-	override	( true | false ) "false" >
-]]>
-
-<!--
-   propval
-
-     This element is for a singly valued property within a property
-     group.  List-valued properties must use the property element above.
-
-     Its attributes are
-
-	name	The name of this property.
-
-	type	The data type for this property.
-
-	value	The value for this property.  Must match type
-		restriction of type attribute.
-
-	override This value should replace any values already in the
-		repository.
--->
-
-<![%profile;[
-<!ELEMENT propval EMPTY>
-
-<!ATTLIST propval
-	name		CDATA #REQUIRED
-	type		( count | integer | opaque | host | hostname |
-			net_address | net_address_v4 | net_address_v6 | time |
-			astring | ustring | boolean | fmri | uri ) #IMPLIED
-	value		CDATA #REQUIRED
-	override	( true | false ) "false" >
-]]>
-
-<![%manifest;[
-<!ELEMENT propval EMPTY>
-
-<!ATTLIST propval
-	name		CDATA #REQUIRED
-	type		( count | integer | opaque | host | hostname |
-			net_address | net_address_v4 | net_address_v6 | time |
-			astring | ustring | boolean | fmri | uri ) #REQUIRED
-	value		CDATA #REQUIRED
-	override	( true | false ) "false" >
-]]>
-
-<!--
-  property_group
-
-    This element is for a set of related properties on a service or
-    instance.  It contains an optional stability element, as well as
-    zero or more property-containing elements.
-
-    Its attributes are
-
-	name	The name of this property group.
-
-	type	A category for this property group.  Groups of type
-		"framework", "implementation" or "template" are primarily
-		of interest to the service management facility, while
-		groups of type "application" are expected to be only of
-		interest to the service to which this group is attached.
-		Other types may be introduced using the service symbol
-		namespace conventions.
-
-	delete	If in the repository, this property group should be removed.
--->
-
-<![%profile;[
-<!ELEMENT property_group
-	( stability?, ( propval | property )* )>
-
-<!ATTLIST property_group
-	name		CDATA #REQUIRED
-	type		CDATA #IMPLIED
-	delete		( true | false ) "false" >
-]]>
-
-<![%manifest;[
-<!ELEMENT property_group
-	( stability?, ( propval | property )* )>
-
-<!ATTLIST property_group
-	name		CDATA #REQUIRED
-	type		CDATA #REQUIRED
-	delete		( true | false ) "false" >
-]]>
-
-<!--
-  service_fmri
-
-    This element defines a reference to a service FMRI (for either a
-    service or an instance).
-
-    Its attribute is
-
-	value	The FMRI.
--->
-
-<!ELEMENT service_fmri EMPTY>
-
-<!ATTLIST service_fmri
-	value		CDATA #REQUIRED>
-
-<!-- Dependencies -->
-
-<!--
-  dependency
-
-    This element identifies a group of FMRIs upon which the service is
-    in some sense dependent.  Its interpretation is left to the
-    restarter to which a particular service instance is delegated.  It
-    contains a group of service FMRIs, as well as a block of properties.
-
-    Its attributes are
-
-	name	The name of this dependency.
-
-	grouping The relationship between the various FMRIs grouped
-		here; "require_all" of the FMRIs to be online, "require_any"
-		of the FMRIs to be online, or "exclude_all" of the FMRIs
-		from being online or in maintenance for the dependency to
-		be satisfied.  "optional_all" dependencies are satisfied
-		when all of the FMRIs are either online or unable to come
-		online (because they are disabled, misconfigured, or one
-		of their dependencies is unable to come online).
-
-	restart_on The type of events from the FMRIs that the service should
-		be restarted for.  "error" restarts the service if the
-		dependency is restarted due to hardware fault.  "restart"
-		restarts the service if the dependency is restarted for
-		any reason, including hardware fault.  "refresh" restarts
-		the service if the dependency is refreshed or restarted for
-		any reason.  "none" will never restart the service due to
-		dependency state changes.
-
-	type	The type of dependency: on another service ('service'), on
-		a filesystem path ('path'), or another dependency type.
-
-	delete	This dependency should be deleted.
--->
-
-<!ELEMENT dependency
-	( service_fmri*, stability?, ( propval | property )* ) >
-
-<!ATTLIST dependency
-	name		CDATA #REQUIRED
-	grouping	( require_all | require_any | exclude_all |
-			optional_all ) #REQUIRED
-	restart_on	( error | restart | refresh | none ) #REQUIRED
-	type		CDATA #REQUIRED
-	delete		( true | false ) "false" >
-
-<!-- Dependents -->
-
-<!--
-  dependent
-
-    This element identifies a service which should depend on this service.  It
-    corresponds to a dependency in the named service.  The grouping and type
-    attributes of that dependency are implied to be "require_all" and
-    "service", respectively.
-
-    Its attributes are
-
-	name	The name of the dependency property group to create in the
-		dependent entity.
-
-	grouping The grouping relationship of the dependency property
-		group to create in the dependent entity.  See "grouping"
-		attribute on the dependency element.
-
-	restart_on The type of events from this service that the named service
-		should be restarted for.
-
-	delete	True if this dependent should be deleted.
-
-	override Whether to replace an existing dependent of the same name.
-
--->
-
-<!ELEMENT dependent
-	( service_fmri, stability?, ( propval | property )* ) >
-
-<!ATTLIST dependent
-	name		CDATA #REQUIRED
-	grouping	( require_all | require_any | exclude_all |
-			optional_all) #REQUIRED
-	restart_on	( error | restart | refresh | none) #REQUIRED
-	delete		( true | false ) "false"
-	override	( true | false ) "false" >
-
-<!-- Method execution context, security profile, and credential definitions -->
-
-<!--
-  envvar
-
-    An environment variable. It has two attributes:
-
-	name	The name of the environment variable.
-	value	The value of the environment variable.
--->
-
-<!ELEMENT envvar EMPTY>
-
-<!ATTLIST envvar
-	name		CDATA #REQUIRED
-	value		CDATA #REQUIRED >
-
-<!--
-  method_environment
-
-    This element defines the environment for a method. It has no
-    attributes, and one or more envvar child elements.
--->
-
-<!ELEMENT method_environment (envvar+) >
-
-<!ATTLIST method_environment>
-
-<!--
-  method_profile
-
-    This element indicates which exec_attr(5) profile applies to the
-    method context being defined.
-
-    Its attribute is
-
-	name	The name of the profile.
--->
-
-<!ELEMENT method_profile EMPTY>
-
-<!ATTLIST method_profile
-	name		CDATA #REQUIRED >
-
-<!--
-  method_credential
-
-    This element specifies credential attributes for the execution
-    method to use.
-
-    Its attributes are
-
-	user	The user ID, in numeric or text form.
-
-	group	The group ID, in numeric or text form.  If absent or
-		":default", the group associated with the user in the
-		passwd database.
-
-	supp_groups Supplementary group IDs to be associated with the
-		method, separated by commas or spaces.  If absent or
-		":default", initgroups(3C) will be used.
-
-	privileges An optional string specifying the privilege set.
-
-	limit_privileges An optional string specifying the limit
-		privilege set.
--->
-
-<!ELEMENT method_credential EMPTY>
-
-<!ATTLIST method_credential
-	user		CDATA #REQUIRED
-	group		CDATA #IMPLIED
-	supp_groups	CDATA #IMPLIED
-	privileges	CDATA #IMPLIED
-	limit_privileges CDATA #IMPLIED >
-
-<!--
-  method_context
-
-    This element combines credential and resource management attributes
-    for execution methods.  It may contain a method_environment, or
-    a method_profile or method_credential element.
-
-    Its attributes are
-
-	working_directory The home directory to launch the method from.
-		":default" can be used as a token to indicate use of the
-		user specified by the credential or profile specified.
-
-	project	The project ID, in numeric or text form.  ":default" can
-		be used as a token to indicate use of the project
-		identified by getdefaultproj(3PROJECT) for the non-root
-		user specified by the credential or profile specified.
-		If the user is root, ":default" designates the project
-		the restarter is running in.
-
-	resource_pool The resource pool name to launch the method on.
-		":default" can be used as a token to indicate use of the
-		pool specified in the project(4) entry given in the
-		"project" attribute above.
--->
-<!ELEMENT method_context
-	( (method_profile | method_credential)?, method_environment? ) >
-
-<!ATTLIST method_context
-	working_directory	CDATA #IMPLIED
-	project			CDATA #IMPLIED
-	resource_pool		CDATA #IMPLIED >
-
-<!-- Restarter delegation, methods, and monitors -->
-
-<!--
-  exec_method
-
-    This element describes one of the methods used by the designated
-    restarter to act on the service instance.  Its interpretation is
-    left to the restarter to which a particular service instance is
-    delegated.  It contains a set of attributes, an optional method
-    context, and an optional stability element for the optional
-    properties that can be included.
-
-    Its attributes are
-
-	type	The type of method, either "method" or "monitor".
-
-	name	Name of this execution method.  The method names are
-		usually a defined interface of the restarter to which an
-		instance of this service is delegated.
-
-	exec	The string identifying the action to take.  For
-		svc.startd(1M), this is a string suitable to pass to
-		exec(2).
-
-	timeout_seconds [integer] Duration, in seconds, to wait for this
-		method to complete.  A '0' or '-1' denotes an infinite
-		timeout.
-
-	delete	If in the repository, the property group for this method
-		should be removed.
--->
-
-<!ELEMENT exec_method
-	( method_context?, stability?, ( propval | property )* ) >
-
-<!ATTLIST exec_method
-	type		( method | monitor ) #REQUIRED
-	name		CDATA #REQUIRED
-	exec		CDATA #REQUIRED
-	timeout_seconds	CDATA #REQUIRED
-	delete		( true | false ) "false" >
-
-<!--
-  restarter
-
-    A flag element identifying the restarter to which this service or
-    service instance is delegated.  Contains the FMRI naming the
-    delegated restarter.
-
-    This element has no attributes.
--->
-
-<!ELEMENT restarter
-	( service_fmri ) >
-
-<!ATTLIST restarter>
-
-<!--
-  Templates
--->
-
-<!--
-  doc_link
-
-    The doc_link relates a resource described by the given URI to the
-    service described by the containing template.  The resource is
-    expected to be a documentation or elucidatory reference of some
-    kind.
-
-    Its attributes are
-
-      name      A label for this resource.
-
-      uri       A URI to the resource.
--->
-
-<!ELEMENT doc_link EMPTY>
-
-<!ATTLIST doc_link
-	name		CDATA #REQUIRED
-	uri		CDATA #REQUIRED >
-
-<!--
-  manpage
-
-    The manpage element connects the reference manual page to the
-    template's service.
-
-    Its attributes are
-
-      title     The manual page title.
-
-      section   The manual page's section.
-
-      manpath   The MANPATH environment variable, as described in man(1)
-                that is required to reach the named manual page
--->
-
-<!ELEMENT manpage EMPTY>
-
-<!ATTLIST manpage
-	title		CDATA #REQUIRED
-	section		CDATA #REQUIRED
-	manpath		CDATA ":default" >
-
-<!--
-  documentation
-
-    The documentation element groups an arbitrary number of doc_link
-    and manpage references.
-
-    It has no attributes.
--->
-
-<!ELEMENT documentation
-	( doc_link | manpage )* >
-
-<!ATTLIST documentation>
-
-<!--
-  loctext
-
-    The loctext element is a container for localized text.
-
-    Its sole attribute is
-
-	xml:lang The name of the locale, in the form accepted by LC_ALL,
-		etc.  See locale(5).
--->
-<!ELEMENT loctext
-        (#PCDATA) >
-
-<!ATTLIST loctext
-        xml:lang	CDATA #REQUIRED >
-
-<!--
-  description
-
-    The description holds a set of potentially longer, localized strings that
-    consist of a short description of the service.
-
-    The description has no attributes.
--->
-<!ELEMENT description
-        ( loctext+ ) >
-
-<!ATTLIST description>
-
-<!--
-  common_name
-
-    The common_name holds a set of short, localized strings that
-    represent a well-known name for the service in the given locale.
-
-    The common_name has no attributes.
--->
-<!ELEMENT common_name
-        ( loctext+ ) >
-
-<!ATTLIST common_name>
-
-<!--
-  units
-
-    The units a numerical property is expressed in.
--->
-
-<!ELEMENT units
-	( loctext+ ) >
-
-<!ATTLIST units>
-
-<!--
-  visibility
-
-    Expresses how a property is typically accessed.  This isn't
-    intended as access control, but as an indicator as to how a
-    property is used.
-
-    Its attributes are:
-
-      value     'hidden', 'readonly', or 'readwrite' indicating that
-		the property should be hidden from the user, shown but
-		read-only, or modifiable.
--->
-
-<!ELEMENT visibility EMPTY>
-
-<!ATTLIST visibility
-	value	( hidden | readonly | readwrite ) #REQUIRED >
-
-<!--
-  value
-
-    Describes a legal value for a property value, and optionally contains a
-    human-readable name and description for the specified property
-    value.
-
-    Its attributes are:
-
-      name	A string representation of the value.
--->
-
-<!ELEMENT value
-	( common_name?, description? ) >
-
-<!ATTLIST value
-	name	CDATA #REQUIRED >
-
-<!--
-  values
-
-    Human-readable names and descriptions for valid values of a property.
--->
-
-<!ELEMENT values
-	(value+) >
-
-<!ATTLIST values>
-
-<!--
-  cardinality
-
-    Places a constraint on the number of values the property can take
-    on.
-
-    Its attributes are:
-	min	minimum number of values
-	max	maximum number of values
-
-    Both attributes are optional.  If min is not specified, it defaults to
-    0.  If max is not specified it indicates an unlimited number of values.
-    If neither is specified this indicates 0 or more values.
--->
-
-<!ELEMENT cardinality EMPTY>
-
-<!ATTLIST cardinality
-	min	CDATA "0"
-	max	CDATA "18446744073709551615">
-
-<!--
-  internal_separators
-
-    Indicates the separators used within a property's value used to
-    separate the actual values.  Used in situations where multiple
-    values are packed into a single property value instead of using a
-    multi-valued property.
--->
-
-<!ELEMENT internal_separators
-	(#PCDATA) >
-
-<!ATTLIST internal_separators>
-
-<!--
-  range
-
-    Indicates a range of possible integer values.
-
-    Its attributes are:
-
-      min	The minimum value of the range (inclusive).
-      max	The maximum value of the range (inclusive).
--->
-
-<!ELEMENT range EMPTY>
-
-<!ATTLIST range
-	min	CDATA #REQUIRED
-	max	CDATA #REQUIRED >
-
-<!--
-  constraints
-
-    Provides a set of constraints on the values a property can take on.
--->
-
-<!ELEMENT constraints
-	( value*, range* ) >
-<!ATTLIST constraints>
-
-<!--
-  include_values
-
-    Includes an entire set of values in the choices block.
-
-    Its attributes are:
-
-	type    Either "constraints" or "values", indicating an
-		inclusion of all values allowed by the property's
-		constraints or all values for which there are
-		human-readable names and descriptions, respectively.
--->
-
-<!ELEMENT include_values EMPTY>
-
-<!ATTLIST include_values
-	type	( constraints | values ) #REQUIRED >
-
-<!--
-  choices
-
-    Provides a set of common choices for the values a property can take
-    on.  Useful in those cases where the possibilities are unenumerable
-    or merely inconveniently legion, and a manageable subset is desired
-    for presentation in a user interface.
--->
-
-<!ELEMENT choices
-	( value*, range*, include_values* ) >
-
-<!ATTLIST choices>
-
-<!--
-  prop_pattern
-
-
-    The prop_pattern describes one property of the enclosing property group
-    pattern.
-
-    Its attributes are:
-
-	name    The property's name.
-	type    The property's type.
-	required
-		If the property group is present, this property is required.
-
-	type can be omitted if required is false.
--->
-
-<!ELEMENT prop_pattern
-	( common_name?, description?, units?, visibility?, cardinality?,
-	  internal_separators?, values?, constraints?, choices? ) >
-
-<!ATTLIST prop_pattern
-	name		CDATA	#REQUIRED
-	type		( count | integer | opaque | host | hostname |
-			net_address | net_address_v4 | net_address_v6 | time |
-			astring | ustring | boolean | fmri | uri ) #IMPLIED
-	required	( true | false )	"false" >
-
-<!--
-  pg_pattern
-
-    The pg_pattern describes one property group.
-    Depending on the element's attributes, these descriptions may apply
-    to just the enclosing service/instance, instances of the enclosing
-    service, delegates of the service (assuming it is a restarter), or
-    all services.
-
-    Its attributes are:
-
-	name    The property group's name.  If not specified, it
-		matches all property groups with the specified type.
-	type    The property group's type.  If not specified, it
-		matches all property groups with the specified name.
-	required
-		If the property group is required.
-	target	The scope of the pattern, which may be all, delegate,
-		instance, or this.  'all' is reserved for framework use
-		and applies the template to all services on the system.
-		'delegate' is reserved for restarters, and means the
-		template applies to all services which use the restarter.
-		'this' would refer to the defining service or instance.
-		'instance' can only be used in a service's template block,
-		and means the definition applies to all instances of this
-		service.
-
--->
-
-<!ELEMENT pg_pattern
-	( common_name?, description?, prop_pattern* ) >
-
-<!ATTLIST pg_pattern
-	name		CDATA	""
-	type		CDATA	""
-	required	( true | false )	"false"
-	target		( this | instance | delegate | all )	"this" >
-
-<!--
-  template
-
-    The template contains a collection of metadata about the service.
-    It contains a localizable string that serves as a common,
-    human-readable name for the service.  (This name should be less than
-    60 characters in a single byte locale.)  The template may optionally
-    contain a longer localizable description of the service, a
-    collection of links to documentation, either in the form of manual
-    pages or in the form of URI specifications to external documentation
-    sources (such as docs.sun.com).
-
-    The template has no attributes.
--->
-<!ELEMENT template
-        ( common_name, description?, documentation?, pg_pattern* ) >
-
-<!ATTLIST template>
-
-<!-- Notification Parameters -->
-
-<!ELEMENT paramval EMPTY>
-
-<!ATTLIST paramval
-	name		CDATA #REQUIRED
-	value		CDATA #REQUIRED>
-
-<!ELEMENT parameter
-	( value_node* )>
-
-<!ATTLIST parameter
-	name		CDATA #REQUIRED>
-
-<!ELEMENT event EMPTY>
-
-<!ATTLIST event
-	value		CDATA #REQUIRED>
-
-<!ELEMENT type
-	( ( parameter | paramval )* )>
-
-<!ATTLIST type
-	name		CDATA #REQUIRED
-	active		( true | false ) "true" >
-
-<!--
-  notification parameters
-
-    This element sets the notification parameters for Software Events and
-    Fault Management problem lifecycle events.
--->
-
-<!ELEMENT notification_parameters
-	( event, type+ )>
-
-<!ATTLIST notification_parameters>
-
-<!-- Services and instances -->
-
-<!--
-  create_default_instance
-
-    A flag element indicating that an otherwise empty default instance
-    of this service (named "default") should be created at install, with
-    its enabled property set as given.
-
-    Its attribute is
-
-	enabled	[boolean] The initial value for the enabled state of
-		this instance.
--->
-
-<!ELEMENT create_default_instance EMPTY >
-
-<!ATTLIST create_default_instance
-	enabled		( true | false ) #REQUIRED >
-
-<!--
-  single_instance
-
-    A flag element stating that this service can only have a single
-    instance on a particular system.
--->
-
-<!ELEMENT single_instance EMPTY>
-
-<!ATTLIST single_instance>
-
-<!--
-  instance
-
-    The service instance is the object representing a software component
-    that will run on the system if enabled.  It contains an enabled
-    element, a set of dependencies on other services, potentially
-    customized methods or configuration data, an optional method
-    context, and a pointer to its restarter.  (If no restarter is
-    specified, the master restarter, svc.startd(1M), is assumed to be
-    responsible for the service.)
-
-    Its attributes are
-
-	name	The canonical name for this instance of the service.
-
-	enabled	[boolean] The initial value for the enabled state of
-		this instance.
--->
-
-<!ELEMENT instance
-	( restarter?, dependency*, dependent*, method_context?,
-	exec_method*, notification_parameters*, property_group*,
-	template? ) >
-
-<!ATTLIST instance
-	name		CDATA #REQUIRED
-	enabled		( true | false ) #REQUIRED >
-
-<!--
-  service
-
-    The service contains the set of instances defined by default for
-    this service, an optional method execution context, any default
-    methods, the template, and various restrictions or advice applicable
-    at installation.  The method execution context and template elements
-    are required for service_bundle documents with type "manifest", but
-    are optional for "profile" or "archive" documents.
-
-    Its attributes are
-
-	name	The canonical name for the service.
-
-	version	[integer] The integer version for this service.
-
-	type	Whether this service is a simple service, a delegated
-		restarter, or a milestone (a synthetic service that
-		collects a group of dependencies).
--->
-
-<!ELEMENT service
-	( create_default_instance?, single_instance?, restarter?,
-	dependency*, dependent*, method_context?, exec_method*,
-	notification_parameters*, property_group*, instance*,
-	stability?, template? ) >
-
-<!ATTLIST service
-	name		CDATA #REQUIRED
-	version		CDATA #REQUIRED
-	type		( service | restarter | milestone ) #REQUIRED >
-
-<!--
-  service_bundle
-
-    The bundle possesses two attributes:
-
-	type	How this file is to be understood by the framework (or
-		used in a non-framework compliant way). Standard types
-		are 'archive', 'manifest', and 'profile'.
-	
-	name	A name for the bundle.  Manifests should be named after
-		the package which delivered them; profiles should be
-		named after the "feature set nickname" they intend to
-		enable.
--->
-
-<!ELEMENT service_bundle
-	( service_bundle* | service* | xi:include* )>
-
-<!ATTLIST service_bundle
-	type		CDATA #REQUIRED
-	name		CDATA #REQUIRED>
